File: attic/belnap_vec_stubs.c
#include <stdint.h>
#include <string.h>

#include <caml/alloc.h>
#include <caml/custom.h>
#include <caml/memory.h>
#include <caml/mlvalues.h>

#if !defined(__has_builtin) || !__has_builtin(__builtin_popcountll)
#    error "__builtin_popcountll is required"
#endif

#if !defined(__has_builtin) || !__has_builtin(__builtin_ctzll)
#    error "__builtin_ctzll is required"
#endif

struct belnap_vec;

#define Bv_val(v) ((struct belnap_vec *)Data_custom_val(v))

/* Must match bits_log2 / bits_mask in belnap_vec.ml. */
enum
{
    BITS_LOG2 = 6,
    BITS_MASK = (1 << BITS_LOG2) - 1
};

static uint64_t const ALL_ONES = ~UINT64_C(0);

/* Layout: words[2*i] = pos plane, words[2*i+1] = neg plane for word-pair i.
   nwords is the per-plane word count = ceil(width/64).
   The flat array has 2*nwords uint64_t elements total. */

struct belnap_vec
{
    int nwords;       /* per-plane word count = ceil(width/64) */
    uint64_t words[]; /* flexible array member; 2*nwords elements */
};

static struct custom_operations bv_ops = {
    "hbt-attic.belnap_vec",
    custom_finalize_default,
    custom_compare_default,
    custom_hash_default,
    custom_serialize_default,
    custom_deserialize_default,
    custom_compare_ext_default,
    custom_fixed_length_default,
};

static inline size_t bv_words_bytes(int nw)
{
    return 2 * (size_t)(nw) * sizeof(uint64_t);
}

CAMLprim value bv_alloc(value vnw)
{
    CAMLparam1(vnw);
    CAMLlocal1(v);
    int const nwords = Int_val(vnw);
    mlsize_t const bsz = sizeof(struct belnap_vec) + bv_words_bytes(nwords);
    v = caml_alloc_custom(&bv_ops, bsz, 0, 1);
    Bv_val(v)->nwords = nwords;
    memset(Bv_val(v)->words, 0, bv_words_bytes(nwords));
    CAMLreturn(v);
}

CAMLprim value bv_blit_grow(value vsrc, value vnew_nw)
{
    CAMLparam2(vsrc, vnew_nw);
    CAMLlocal1(vdst);
    int const new_nwords = Int_val(vnew_nw);
    mlsize_t const bsz = sizeof(struct belnap_vec) + bv_words_bytes(new_nwords);
    vdst = caml_alloc_custom(&bv_ops, bsz, 0, 1);
    struct belnap_vec *dst = Bv_val(vdst);
    struct belnap_vec const *src = Bv_val(vsrc);
    int const old_nwords = src->nwords;
    dst->nwords = new_nwords;
    memcpy(dst->words, src->words, bv_words_bytes(old_nwords));
    memset(dst->words + (2 * old_nwords), 0, bv_words_bytes(new_nwords - old_nwords));
    CAMLreturn(vdst);
}

CAMLprim value bv_get(value vbv, value vi)
{
    CAMLparam2(vbv, vi);
    uint64_t const *const w = Bv_val(vbv)->words;
    int const i = Int_val(vi);

    int const word = i >> BITS_LOG2;
    int const bit = i & BITS_MASK;
    int const pos_bit = (int)((w[word * 2] >> bit) & 1);
    int const neg_bit = (int)((w[word * 2 + 1] >> bit) & 1);
    CAMLreturn(Val_int((neg_bit << 1) | pos_bit));
}

CAMLprim value bv_set(value vbv, value vi, value vraw)
{
    CAMLparam3(vbv, vi, vraw);
    uint64_t *const w = Bv_val(vbv)->words;
    int const i = Int_val(vi);
    int const raw = Int_val(vraw);

    int const word = i >> BITS_LOG2;
    int const bit = i & BITS_MASK;
    uint64_t const mask = UINT64_C(1) << bit;
    uint64_t const pos = (uint64_t)(raw & 1) << bit;
    uint64_t const neg = (uint64_t)((raw >> 1) & 1) << bit;
    w[word * 2] = (w[word * 2] & ~mask) | pos;
    w[word * 2 + 1] = (w[word * 2 + 1] & ~mask) | neg;
    CAMLreturn(Val_unit);
}

CAMLprim value bv_mask_tail(value vbv, value vwidth)
{
    CAMLparam2(vbv, vwidth);
    uint64_t *const w = Bv_val(vbv)->words;
    int const nw = Bv_val(vbv)->nwords;
    int const width = Int_val(vwidth);

    int const r = width & BITS_MASK;
    if (nw > 0 && r != 0)
    {
        uint64_t const mask = (UINT64_C(1) << r) - 1;
        int const base = (nw - 1) * 2;
        w[base] &= mask;
        w[base + 1] &= mask;
    }
    CAMLreturn(Val_unit);
}

CAMLprim value bv_fill(value vbv, value vfrom, value vto, value vraw)
{
    CAMLparam4(vbv, vfrom, vto, vraw);
    uint64_t *const w = Bv_val(vbv)->words;
    int const from = Int_val(vfrom);
    int const to_ = Int_val(vto);
    int const raw = Int_val(vraw);

    uint64_t const pos = (raw & 1) ? ALL_ONES : UINT64_C(0);
    uint64_t const neg = (raw >> 1) & 1 ? ALL_ONES : UINT64_C(0);

    for (int i = from; i < to_; i++)
    {
        w[i * 2] = pos;
        w[i * 2 + 1] = neg;
    }

    CAMLreturn(Val_unit);
}

CAMLprim value bv_not(value src, value dst)
{
    CAMLparam2(src, dst);
    uint64_t const *const s = Bv_val(src)->words;
    uint64_t *const d = Bv_val(dst)->words;
    int const nw = Bv_val(src)->nwords;

    for (int i = 0; i < nw; i++)
    {
        d[i * 2] = s[i * 2 + 1];
        d[i * 2 + 1] = s[i * 2];
    }

    CAMLreturn(Val_unit);
}

CAMLprim value bv_and(value va, value vb, value dst)
{
    CAMLparam3(va, vb, dst);
    uint64_t const *const a = Bv_val(va)->words;
    uint64_t const *const b = Bv_val(vb)->words;
    uint64_t *const d = Bv_val(dst)->words;
    int const nwa = Bv_val(va)->nwords;
    int const nwb = Bv_val(vb)->nwords;

    int const nw = nwa > nwb ? nwa : nwb;

    for (int i = 0; i < nw; i++)
    {
        uint64_t const ap = i < nwa ? a[i * 2] : 0;
        uint64_t const an = i < nwa ? a[i * 2 + 1] : 0;
        uint64_t const bp = i < nwb ? b[i * 2] : 0;
        uint64_t const bn = i < nwb ? b[i * 2 + 1] : 0;
        d[i * 2] = ap & bp;
        d[i * 2 + 1] = an | bn;
    }

    CAMLreturn(Val_unit);
}

CAMLprim value bv_or(value va, value vb, value dst)
{
    CAMLparam3(va, vb, dst);
    uint64_t const *const a = Bv_val(va)->words;
    uint64_t const *const b = Bv_val(vb)->words;
    uint64_t *const d = Bv_val(dst)->words;
    int const nwa = Bv_val(va)->nwords;
    int const nwb = Bv_val(vb)->nwords;

    int const nw = nwa > nwb ? nwa : nwb;

    for (int i = 0; i < nw; i++)
    {
        uint64_t const ap = i < nwa ? a[i * 2] : 0;
        uint64_t const an = i < nwa ? a[i * 2 + 1] : 0;
        uint64_t const bp = i < nwb ? b[i * 2] : 0;
        uint64_t const bn = i < nwb ? b[i * 2 + 1] : 0;
        d[i * 2] = ap | bp;
        d[i * 2 + 1] = an & bn;
    }

    CAMLreturn(Val_unit);
}

CAMLprim value bv_merge(value va, value vb, value dst)
{
    CAMLparam3(va, vb, dst);
    uint64_t const *const a = Bv_val(va)->words;
    uint64_t const *const b = Bv_val(vb)->words;
    uint64_t *const d = Bv_val(dst)->words;
    int const nwa = Bv_val(va)->nwords;
    int const nwb = Bv_val(vb)->nwords;

    int const nw = nwa > nwb ? nwa : nwb;

    for (int i = 0; i < nw; i++)
    {
        uint64_t const ap = i < nwa ? a[i * 2] : 0;
        uint64_t const an = i < nwa ? a[i * 2 + 1] : 0;
        uint64_t const bp = i < nwb ? b[i * 2] : 0;
        uint64_t const bn = i < nwb ? b[i * 2 + 1] : 0;
        d[i * 2] = ap | bp;
        d[i * 2 + 1] = an | bn;
    }

    CAMLreturn(Val_unit);
}

CAMLprim value bv_consensus(value va, value vb, value dst)
{
    CAMLparam3(va, vb, dst);
    uint64_t const *const a = Bv_val(va)->words;
    uint64_t const *const b = Bv_val(vb)->words;
    uint64_t *const d = Bv_val(dst)->words;
    int const nwa = Bv_val(va)->nwords;
    int const nwb = Bv_val(vb)->nwords;

    int const nw = nwa > nwb ? nwa : nwb;

    for (int i = 0; i < nw; i++)
    {
        uint64_t const ap = i < nwa ? a[i * 2] : 0;
        uint64_t const an = i < nwa ? a[i * 2 + 1] : 0;
        uint64_t const bp = i < nwb ? b[i * 2] : 0;
        uint64_t const bn = i < nwb ? b[i * 2 + 1] : 0;
        d[i * 2] = ap & bp;        /* pos plane: AND */
        d[i * 2 + 1] = an & bn;   /* neg plane: AND */
    }

    CAMLreturn(Val_unit);
}

static inline uint64_t tail_mask(int const width)
{
    int const r = width & BITS_MASK;
    return r == 0 ? ALL_ONES : (UINT64_C(1) << r) - 1;
}

CAMLprim value bv_is_consistent(value vbv, value vwidth)
{
    CAMLparam2(vbv, vwidth);
    uint64_t const *const w = Bv_val(vbv)->words;
    int const nw = Bv_val(vbv)->nwords;
    int const width = Int_val(vwidth);

    for (int i = 0; i < nw; i++)
    {
        uint64_t const m = (i == nw - 1) ? tail_mask(width) : ALL_ONES;
        if ((w[i * 2] & w[i * 2 + 1] & m) != 0)
            CAMLreturn(Val_int(0));
    }

    CAMLreturn(Val_int(1));
}

CAMLprim value bv_is_all_determined(value vbv, value vwidth)
{
    CAMLparam2(vbv, vwidth);
    uint64_t const *const w = Bv_val(vbv)->words;
    int const nw = Bv_val(vbv)->nwords;
    int const width = Int_val(vwidth);

    for (int i = 0; i < nw; i++)
    {
        uint64_t const m = (i == nw - 1) ? tail_mask(width) : ALL_ONES;
        uint64_t const xorv = w[i * 2] ^ w[i * 2 + 1];
        if ((xorv & m) != m)
            CAMLreturn(Val_int(0));
    }

    CAMLreturn(Val_int(1));
}

CAMLprim value bv_is_all_true(value vbv, value vwidth)
{
    CAMLparam2(vbv, vwidth);
    uint64_t const *const w = Bv_val(vbv)->words;
    int const nw = Bv_val(vbv)->nwords;
    int const width = Int_val(vwidth);

    for (int i = 0; i < nw; i++)
    {
        uint64_t const m = (i == nw - 1) ? tail_mask(width) : ALL_ONES;
        if ((w[i * 2] & m) != m)
            CAMLreturn(Val_int(0)); /* not all pos set */
        if ((w[i * 2 + 1] & m) != 0)
            CAMLreturn(Val_int(0)); /* some neg set */
    }

    CAMLreturn(Val_int(1));
}

CAMLprim value bv_is_all_false(value vbv, value vwidth)
{
    CAMLparam2(vbv, vwidth);
    uint64_t const *const w = Bv_val(vbv)->words;
    int const nw = Bv_val(vbv)->nwords;
    int const width = Int_val(vwidth);

    for (int i = 0; i < nw; i++)
    {
        uint64_t const m = (i == nw - 1) ? tail_mask(width) : ALL_ONES;
        if ((w[i * 2] & m) != 0)
            CAMLreturn(Val_int(0)); /* some pos set */
        if ((w[i * 2 + 1] & m) != m)
            CAMLreturn(Val_int(0)); /* not all neg set */
    }

    CAMLreturn(Val_int(1));
}

static inline int popcount64(uint64_t x)
{
    return __builtin_popcountll(x);
}

CAMLprim value bv_count_true(value vbv)
{
    CAMLparam1(vbv);
    uint64_t const *const w = Bv_val(vbv)->words;
    int const nw = Bv_val(vbv)->nwords;

    int n = 0;
    for (int i = 0; i < nw; i++)
        n += popcount64(w[i * 2] & ~w[i * 2 + 1]);
    CAMLreturn(Val_int(n));
}

CAMLprim value bv_count_false(value vbv)
{
    CAMLparam1(vbv);
    uint64_t const *const w = Bv_val(vbv)->words;
    int const nw = Bv_val(vbv)->nwords;

    int n = 0;
    for (int i = 0; i < nw; i++)
        n += popcount64(~w[i * 2] & w[i * 2 + 1]);
    CAMLreturn(Val_int(n));
}

CAMLprim value bv_count_both(value vbv)
{
    CAMLparam1(vbv);
    uint64_t const *const w = Bv_val(vbv)->words;
    int const nw = Bv_val(vbv)->nwords;

    int n = 0;
    for (int i = 0; i < nw; i++)
        n += popcount64(w[i * 2] & w[i * 2 + 1]);
    CAMLreturn(Val_int(n));
}

CAMLprim value bv_equal(value va, value vb)
{
    CAMLparam2(va, vb);
    struct belnap_vec const *a = Bv_val(va);
    struct belnap_vec const *b = Bv_val(vb);
    if (a->nwords != b->nwords)
        CAMLreturn(Val_int(0));
    CAMLreturn(Val_int(memcmp(a->words, b->words, bv_words_bytes(a->nwords)) == 0));
}

CAMLprim value bv_init_from_list(value vbv, value vlist)
{
    CAMLparam2(vbv, vlist);
    struct belnap_vec *bv = Bv_val(vbv);
    int i = 0;
    uint64_t pos = 0, neg = 0;
    while (vlist != Val_emptylist)
    {
        int const raw = Int_val(Field(vlist, 0));
        int const bit = i & BITS_MASK;
        pos |= (uint64_t)(raw & 1) << bit;
        neg |= (uint64_t)((raw >> 1) & 1) << bit;
        vlist = Field(vlist, 1);
        if (bit == BITS_MASK || vlist == Val_emptylist)
        {
            int const word = i >> BITS_LOG2;
            bv->words[word * 2] = pos;
            bv->words[word * 2 + 1] = neg;
            pos = 0;
            neg = 0;
        }
        i++;
    }
    CAMLreturn(Val_unit);
}

CAMLprim value bv_init_from_array(value vbv, value varr)
{
    CAMLparam2(vbv, varr);
    struct belnap_vec *bv = Bv_val(vbv);
    int const width = (int)Wosize_val(varr);
    uint64_t pos = 0, neg = 0;
    for (int i = 0; i < width; i++)
    {
        int const raw = Int_val(Field(varr, i));
        int const bit = i & BITS_MASK;
        pos |= (uint64_t)(raw & 1) << bit;
        neg |= (uint64_t)((raw >> 1) & 1) << bit;
        if (bit == BITS_MASK || i == width - 1)
        {
            int const word = i >> BITS_LOG2;
            bv->words[word * 2] = pos;
            bv->words[word * 2 + 1] = neg;
            pos = 0;
            neg = 0;
        }
    }
    CAMLreturn(Val_unit);
}

CAMLprim value bv_to_array(value vbv, value vwidth)
{
    CAMLparam2(vbv, vwidth);
    CAMLlocal1(arr);
    int const width = Int_val(vwidth);
    if (width == 0)
        CAMLreturn(Atom(0));
    arr = caml_alloc(width, 0);
    /* Extract raw pointers after allocation so they are not stale if GC ran. */
    struct belnap_vec const *const bv = Bv_val(vbv);
    int const nwords = bv->nwords;
    uint64_t const *const w = bv->words;
    /* No allocation below; Store_field write-barrier is a no-op for immediates. */
    for (int i = 0; i < nwords; i++)
    {
        uint64_t const pos = w[i * 2];
        uint64_t const neg = w[i * 2 + 1];
        int const base = i << BITS_LOG2;
        int const limit = base + 64 < width ? 64 : width - base;
        for (int bit = 0; bit < limit; bit++)
        {
            int const raw = (int)(((pos >> bit) & 1) | (((neg >> bit) & 1) << 1));
            Store_field(arr, base + bit, Val_int(raw));
        }
    }
    CAMLreturn(arr);
}

CAMLprim value bv_find_first(value vbv, value vwidth, value vraw)
{
    CAMLparam3(vbv, vwidth, vraw);
    uint64_t const *const w = Bv_val(vbv)->words;
    int const nw = Bv_val(vbv)->nwords;
    int const width = Int_val(vwidth);
    int const raw = Int_val(vraw);
    uint64_t const want_pos = (uint64_t)(raw & 1);
    uint64_t const want_neg = (uint64_t)((raw >> 1) & 1);

    for (int i = 0; i < nw; i++)
    {
        uint64_t const m = (i == nw - 1) ? tail_mask(width) : ALL_ONES;
        uint64_t const pos_word = w[i * 2];
        uint64_t const neg_word = w[i * 2 + 1];
        /* Match positions where pos == want_pos AND neg == want_neg. */
        uint64_t const pos_match = want_pos ? pos_word : ~pos_word;
        uint64_t const neg_match = want_neg ? neg_word : ~neg_word;
        uint64_t const candidates = pos_match & neg_match & m;
        if (candidates != 0)
            CAMLreturn(Val_int((i << BITS_LOG2) + __builtin_ctzll(candidates)));
    }
    CAMLreturn(Val_int(-1));
}

File: attic/belnap_vec_stubs.js
// belnap_vec_stubs.js
//
// js_of_ocaml JavaScript stubs for Belnap vector operations.
//
// Each bv is a plain JS object { nwords, words } where words is a
// BigUint64Array of length 2*nwords in interleaved pos/neg layout:
//   words[2*i] = pos-plane word i, words[2*i+1] = neg-plane word i.
//
// OCaml integers arrive and depart as plain JS numbers (no tagging).
// OCaml unit is 0.  OCaml bool/int results are plain 0/1.
//
// BigInt bitwise NOT (~x) is unbounded two's-complement.  ANDing ~x
// with a BigUint64Array element (at most 64 bits) bounds the result
// naturally, so explicit masking after NOT is only needed when both
// operands could be negative (see bv_find_first).

//Provides: bv_all_ones
var bv_all_ones = 0xffffffffffffffffn;

//Provides: bv_tail_mask
//Requires: bv_all_ones
function bv_tail_mask(width) {
    var r = width & 63;
    return r === 0 ? bv_all_ones : (1n << BigInt(r)) - 1n;
}

//Provides: bv_popcount64
function bv_popcount64(v) {
    var n = 0;
    while (v !== 0n) {
        v &= v - 1n;
        n++;
    }
    return n;
}

//Provides: bv_ctz64
function bv_ctz64(v) {
    var n = 0;
    while ((v & 1n) === 0n) {
        v >>= 1n;
        n++;
    }
    return n;
}

//Provides: bv_alloc
function bv_alloc(nw) {
    return { nwords: nw, words: new globalThis.BigUint64Array(2 * nw) };
}

//Provides: bv_blit_grow
//Requires: bv_alloc
function bv_blit_grow(src, new_nw) {
    var dst = bv_alloc(new_nw);
    var old_nw = src.nwords;
    for (var i = 0; i < 2 * old_nw; i++) {
        dst.words[i] = src.words[i];
    }
    return dst;
}

//Provides: bv_get
function bv_get(bv, i) {
    var word = i >> 6;
    var bit = BigInt(i & 63);
    var pos_bit = Number((bv.words[word * 2] >> bit) & 1n);
    var neg_bit = Number((bv.words[word * 2 + 1] >> bit) & 1n);
    return (neg_bit << 1) | pos_bit;
}

//Provides: bv_set
function bv_set(bv, i, raw) {
    var word = i >> 6;
    var bit = BigInt(i & 63);
    var mask = 1n << bit;
    var pos = BigInt(raw & 1) << bit;
    var neg = BigInt((raw >> 1) & 1) << bit;
    bv.words[word * 2] = (bv.words[word * 2] & ~mask) | pos;
    bv.words[word * 2 + 1] = (bv.words[word * 2 + 1] & ~mask) | neg;
    return 0;
}

//Provides: bv_mask_tail
function bv_mask_tail(bv, width) {
    var nw = bv.nwords;
    var r = width & 63;
    if (nw > 0 && r !== 0) {
        var mask = (1n << BigInt(r)) - 1n;
        var base = (nw - 1) * 2;
        bv.words[base] &= mask;
        bv.words[base + 1] &= mask;
    }
    return 0;
}

//Provides: bv_fill
//Requires: bv_all_ones
function bv_fill(bv, from, to, raw) {
    var pos = raw & 1 ? bv_all_ones : 0n;
    var neg = (raw >> 1) & 1 ? bv_all_ones : 0n;
    for (var i = from; i < to; i++) {
        bv.words[i * 2] = pos;
        bv.words[i * 2 + 1] = neg;
    }
    return 0;
}

//Provides: bv_not
function bv_not(src, dst) {
    var nw = src.nwords;
    for (var i = 0; i < nw; i++) {
        dst.words[i * 2] = src.words[i * 2 + 1];
        dst.words[i * 2 + 1] = src.words[i * 2];
    }
    return 0;
}

//Provides: bv_and
function bv_and(a, b, dst) {
    var nw = a.nwords > b.nwords ? a.nwords : b.nwords;
    for (var i = 0; i < nw; i++) {
        var ap = i < a.nwords ? a.words[i * 2] : 0n;
        var an = i < a.nwords ? a.words[i * 2 + 1] : 0n;
        var bp = i < b.nwords ? b.words[i * 2] : 0n;
        var bn = i < b.nwords ? b.words[i * 2 + 1] : 0n;
        dst.words[i * 2] = ap & bp;
        dst.words[i * 2 + 1] = an | bn;
    }
    return 0;
}

//Provides: bv_or
function bv_or(a, b, dst) {
    var nw = a.nwords > b.nwords ? a.nwords : b.nwords;
    for (var i = 0; i < nw; i++) {
        var ap = i < a.nwords ? a.words[i * 2] : 0n;
        var an = i < a.nwords ? a.words[i * 2 + 1] : 0n;
        var bp = i < b.nwords ? b.words[i * 2] : 0n;
        var bn = i < b.nwords ? b.words[i * 2 + 1] : 0n;
        dst.words[i * 2] = ap | bp;
        dst.words[i * 2 + 1] = an & bn;
    }
    return 0;
}

//Provides: bv_merge
function bv_merge(a, b, dst) {
    var nw = a.nwords > b.nwords ? a.nwords : b.nwords;
    for (var i = 0; i < nw; i++) {
        var ap = i < a.nwords ? a.words[i * 2] : 0n;
        var an = i < a.nwords ? a.words[i * 2 + 1] : 0n;
        var bp = i < b.nwords ? b.words[i * 2] : 0n;
        var bn = i < b.nwords ? b.words[i * 2 + 1] : 0n;
        dst.words[i * 2] = ap | bp;
        dst.words[i * 2 + 1] = an | bn;
    }
    return 0;
}

//Provides: bv_is_consistent
//Requires: bv_all_ones, bv_tail_mask
function bv_is_consistent(bv, width) {
    var nw = bv.nwords;
    for (var i = 0; i < nw; i++) {
        var m = i === nw - 1 ? bv_tail_mask(width) : bv_all_ones;
        if ((bv.words[i * 2] & bv.words[i * 2 + 1] & m) !== 0n) {
            return 0;
        }
    }
    return 1;
}

//Provides: bv_is_all_determined
//Requires: bv_all_ones, bv_tail_mask
function bv_is_all_determined(bv, width) {
    var nw = bv.nwords;
    for (var i = 0; i < nw; i++) {
        var m = i === nw - 1 ? bv_tail_mask(width) : bv_all_ones;
        var xorv = bv.words[i * 2] ^ bv.words[i * 2 + 1];
        if ((xorv & m) !== m) {
            return 0;
        }
    }
    return 1;
}

//Provides: bv_is_all_true
//Requires: bv_all_ones, bv_tail_mask
function bv_is_all_true(bv, width) {
    var nw = bv.nwords;
    for (var i = 0; i < nw; i++) {
        var m = i === nw - 1 ? bv_tail_mask(width) : bv_all_ones;
        if ((bv.words[i * 2] & m) !== m) {
            // not all pos set
            return 0;
        }
        if ((bv.words[i * 2 + 1] & m) !== 0n) {
            // some neg set
            return 0;
        }
    }
    return 1;
}

//Provides: bv_is_all_false
//Requires: bv_all_ones, bv_tail_mask
function bv_is_all_false(bv, width) {
    var nw = bv.nwords;
    for (var i = 0; i < nw; i++) {
        var m = i === nw - 1 ? bv_tail_mask(width) : bv_all_ones;
        if ((bv.words[i * 2] & m) !== 0n) {
            // some pos set
            return 0;
        }
        if ((bv.words[i * 2 + 1] & m) !== m) {
            // not all neg set
            return 0;
        }
    }
    return 1;
}

//Provides: bv_count_true
//Requires: bv_popcount64
function bv_count_true(bv) {
    var n = 0;
    for (var i = 0; i < bv.nwords; i++) {
        n += bv_popcount64(bv.words[i * 2] & ~bv.words[i * 2 + 1]);
    }
    return n;
}

//Provides: bv_count_false
//Requires: bv_popcount64
function bv_count_false(bv) {
    var n = 0;
    for (var i = 0; i < bv.nwords; i++) {
        n += bv_popcount64(~bv.words[i * 2] & bv.words[i * 2 + 1]);
    }
    return n;
}

//Provides: bv_count_both
//Requires: bv_popcount64
function bv_count_both(bv) {
    var n = 0;
    for (var i = 0; i < bv.nwords; i++) {
        n += bv_popcount64(bv.words[i * 2] & bv.words[i * 2 + 1]);
    }
    return n;
}

//Provides: bv_equal
function bv_equal(a, b) {
    if (a.nwords !== b.nwords) return 0;
    for (var i = 0; i < 2 * a.nwords; i++) {
        if (a.words[i] !== b.words[i]) {
            return 0;
        }
    }
    return 1;
}

//Provides: bv_init_from_list
function bv_init_from_list(bv, list) {
    var i = 0,
        pos = 0n,
        neg = 0n;
    // 0 = Val_emptylist in jsoo
    while (list !== 0) {
        // Field(list, 0): head
        var raw = list[1];
        // Field(list, 1): tail
        list = list[2];
        var bit = BigInt(i & 63);
        pos |= BigInt(raw & 1) << bit;
        neg |= BigInt((raw >> 1) & 1) << bit;
        if ((i & 63) === 63 || list === 0) {
            var word = i >> 6;
            bv.words[word * 2] = pos;
            bv.words[word * 2 + 1] = neg;
            pos = 0n;
            neg = 0n;
        }
        i++;
    }
    return 0;
}

//Provides: bv_init_from_array
function bv_init_from_array(bv, arr) {
    // Wosize_val: arr = [tag, v0, …]
    var width = arr.length - 1;
    var pos = 0n,
        neg = 0n;
    for (var i = 0; i < width; i++) {
        // Field(arr, i)
        var raw = arr[i + 1];
        var bit = BigInt(i & 63);
        pos |= BigInt(raw & 1) << bit;
        neg |= BigInt((raw >> 1) & 1) << bit;
        if ((i & 63) === 63 || i === width - 1) {
            var word = i >> 6;
            bv.words[word * 2] = pos;
            bv.words[word * 2 + 1] = neg;
            pos = 0n;
            neg = 0n;
        }
    }
    return 0;
}

//Provides: bv_to_array
function bv_to_array(bv, width) {
    // Atom(0): empty array block
    if (width === 0) return [0];
    var arr = new Array(width + 1);
    // block tag
    arr[0] = 0;
    var nw = bv.nwords;
    for (var i = 0; i < nw; i++) {
        var pos = bv.words[i * 2];
        var neg = bv.words[i * 2 + 1];
        var base = i << 6;
        var limit = width - base < 64 ? width - base : 64;
        for (var bit = 0; bit < limit; bit++) {
            var b = BigInt(bit);
            var raw = Number((pos >> b) & 1n) | (Number((neg >> b) & 1n) << 1);
            // Store_field(arr, base+bit, …)
            arr[base + bit + 1] = raw;
        }
    }
    return arr;
}

//Provides: bv_find_first
//Requires: bv_all_ones, bv_tail_mask, bv_ctz64
function bv_find_first(bv, width, raw) {
    var nw = bv.nwords;
    var want_pos = BigInt(raw & 1);
    var want_neg = BigInt((raw >> 1) & 1);
    for (var i = 0; i < nw; i++) {
        var m = i === nw - 1 ? bv_tail_mask(width) : bv_all_ones;
        var pos_word = bv.words[i * 2];
        var neg_word = bv.words[i * 2 + 1];
        var pos_match = want_pos ? pos_word : ~pos_word;
        var neg_match = want_neg ? neg_word : ~neg_word;
        var candidates = pos_match & neg_match & m;
        if (candidates !== 0n) {
            return (i << 6) + bv_ctz64(candidates);
        }
    }
    return -1;
}
